quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,71 +198,179 @@ impl Recv {
 
-pub(super) type StreamReadResult<T> = Result<Option<T>, ReadError>;
+impl<'a> Chunks<'a> {
+    pub(crate) fn new(
+        id: StreamId,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+        ordered: bool,
+    ) -> Result<Self, ReadableError> {
+        let entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadableError::UnknownStream),
+        };
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
-}
+        let mut rs = match entry.get().stopped {
+            true => return Err(ReadableError::UnknownStream),
+            false => entry.remove(),
+        };
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+        rs.assembler.ensure_ordering(ordered)?;
+        Ok(Self {
+            id,
+            recv: Some(rs),
+            streams,
+            pending,
+            ordered,
+            read: 0,
+            fused: None,
+        })
     }
-}
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+    /// Read the next chunk from the stream
+    ///
+    /// Return the next chunk of data from the stream, if any are available. `max_length` may be
+    /// used to restrict the maximum size of the returned `Bytes` (but `usize::MAX` is supported,
+    /// too).
+    pub fn next(&mut self, max_length: usize) -> ReadResult {
+        if let Some(status) = &self.fused {
+            return ReadResult::from_status(*status);
+        }
+
+        let rs = self.recv.as_mut().unwrap();
+        let chunk = rs.assembler.read(max_length, self.ordered);
+        let status = match &chunk {
+            Some(chunk) => {
+                self.read += chunk.bytes.len() as u64;
+                ReadStatus::Readable
+            }
+            None => match rs.state {
+                RecvState::Recv { size } => {
+                    if size == Some(rs.end) && rs.assembler.bytes_read() == rs.end {
+                        self.streams.stream_freed(self.id, StreamHalf::Recv);
+                        self.fused = Some(ReadStatus::Finished);
+                        ReadStatus::Finished
+                    } else {
+                        self.streams.recv.insert(self.id, self.recv.take().unwrap());
+                        self.fused = Some(ReadStatus::Blocked);
+                        ReadStatus::Blocked
+                    }
+                }
+                RecvState::ResetRecvd { error_code, .. } => {
+                    rs.state = RecvState::Closed;
+                    self.streams.stream_freed(self.id, StreamHalf::Recv);
+                    self.fused = Some(ReadStatus::Reset(error_code));
+                    ReadStatus::Reset(error_code)
+                }
+                RecvState::Closed => unreachable!(),
+            },
+        };
+
+        let mut should_transmit = ShouldTransmit(false);
+        if status != ReadStatus::Readable {
```

Comment:
Is this condition right? if we have a large receive buffer which is full, and the user only reads 25% of it, we would need to send an update. But the state would stay at readable.

Reply:
I don't think it would be efficient to suggest to the caller that they `poll_transmit()` if there is yet more data to read, right? They should ideally read out all of the available data and then inspect `ShouldTransmit` to see if polling for packets to be sent is necessary.

Reply:
There is no need or guarantee that a caller will read all available data. E.g. a HTTP/3 server might just enough chunks until it determines it got the header of a request. As soon as it has the header it might execute some business logic - e.g. connecting to a upstream service, preparing a response, etc. That might take hundreds of ms, and for that time it won't bother reading more buffered data. However this shouldn't block the peer from at least sending some more data.

It's the same as there is no expectation that a user of a TCP socket immediately drains it completely, instead of letting data being buffered in there.

Reply:
I completely rewrote this part to be much more like the current logic.